### PR TITLE
Add tests in System.ComponentModel.Annotations

### PR DESCRIPTION
**Pull request by hughbe**
And remove some dead code
/cc @stephentoub

### CODE REVIEW

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttribute.cs [hunk 1]
```diff
@@ -71,18 +72,14 @@ protected ValidationAttribute(Func<string> errorMessageAccessor)
 
         #region Internal Properties
         /// <summary>
-        /// Gets or sets and the default error message string.
+        /// Sets and the default error message string.
```
Comment (stephentoub) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttribute.cs]: Nit: preexisting your change, but the ```and``` here is erroneous

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttribute.cs [hunk 2]
```diff
@@ -252,54 +249,48 @@ private void SetupResourceAccessor()
...
+                        CultureInfo.CurrentCulture,
+                        SR.ValidationAttribute_ResourceTypeDoesNotHaveProperty,
+                        _errorMessageResourceType.FullName,
+                        _errorMessageResourceName));
             }
-            else
+
+            if (property.PropertyType != typeof(string))
             {
-                throw new InvalidOperationException(string.Format(CultureInfo.CurrentCulture,
```
Comment (stephentoub) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttribute.cs]: How did you determine that this is dead code?

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs [hunk 3]
```diff
@@ -216,10 +214,7 @@ internal bool TryGetPropertyStoreItem(string propertyName, out PropertyStoreItem
                 {
                     lock (_syncRoot)
                     {
-                        if (_propertyStoreItems == null)
```
Comment (stephentoub) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs]: Why did you delete this check?  Isn't it necessary for thread-safety as part of the double-checked locking scheme bein employed?

File: src/System.ComponentModel.Annotations/tests/FilterUIHintAttributeTests.cs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using Xunit;
+
+namespace System.ComponentModel.DataAnnotations.Tests
+{
+    public class FilterUIHintAttributeTests
+    {
+#pragma warning disable 618
```
Comment (stephentoub) [inline src/System.ComponentModel.Annotations/tests/FilterUIHintAttributeTests.cs]: Can you add a comment next to this about why it's necessary?  Presumably it's to suppress the obsolete warnings for FilterUIHintAttribute usage?

File: src/System.ComponentModel.Annotations/tests/PhoneAttributeTests.cs [hunk 5]
```diff
@@ -23,6 +23,13 @@ protected override IEnumerable<TestCase> ValidValues()
             yield return new TestCase(new PhoneAttribute(), "425-555-1212 ext 123");
             yield return new TestCase(new PhoneAttribute(), "425-555-1212 ext.123");
             yield return new TestCase(new PhoneAttribute(), "425-555-1212 ext. 123");
+            yield return new TestCase(new PhoneAttribute(), "1");
+            yield return new TestCase(new PhoneAttribute(), "+4+2+5+-+5+5+5+-+1+2++1+2++");
+            yield return new TestCase(new PhoneAttribute(), "425-555-1212    ");
+            yield return new TestCase(new PhoneAttribute(), " \r \n 1  \t ");
+            yield return new TestCase(new PhoneAttribute(), "1-.()");
+            yield return new TestCase(new PhoneAttribute(), "(425555-1212");
+            yield return new TestCase(new PhoneAttribute(), ")425555-1212");
```
Comment (stephentoub) [inline src/System.ComponentModel.Annotations/tests/PhoneAttributeTests.cs]: Just curious, are you adding these based on seeing what code paths are missing in the code coverage report, or something else?

**Comment by stephentoub** [comment]
The tests look good; thanks for adding those.  I have some questions on the src changes.

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttribute.cs (hunk 2 above)
Comment (hughbe) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttribute.cs]: This method is called from the method SetupResourceAccessor(), which does the following:
```
private void SetupResourceAccessor()
{
	bool resourceNameSet = !string.IsNullOrEmpty(_errorMessageResourceName);
	bool resourceTypeSet = _errorMessageResourceType != null;
	// ...
	if (resourceNameSet != resourceTypeSet) throw new InvalidOperationException();
	if (resourceNameSet)
	{
	    SetResourceAccessorByPropertyLookup();
	}
    else 
    {
        // ...
    }
}
```
This is the only point that `SetResourceAccessorByPropertyLookup` is called.
If resourceNameSet is true, then resourceTypeSet has to be true when we call the method, as we throw an InvalidOperationException if this is not the case.
Thus, when we call `SetResourceAccessorByPropertyLookup`, and check `if (_errorMessageResourceType != null && !string.IsNullOrEmpty(_errorMessageResourceName))`, this case will always be true, as:
```
resourceTypeSet == true, therefore _errorMessageResourceType != null
resourceNameSet == true, therefore !string.IsNullOrEmpty(_errorMessageResourceName)
```

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs (hunk 3 above)
Comment (hughbe) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs]: Had no idea about this paradigm, added back

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttribute.cs (hunk 1 above)
Comment (hughbe) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttribute.cs]: fixed

File: src/System.ComponentModel.Annotations/tests/FilterUIHintAttributeTests.cs (hunk 4 above)
Comment (hughbe) [inline src/System.ComponentModel.Annotations/tests/FilterUIHintAttributeTests.cs]: done

File: src/System.ComponentModel.Annotations/tests/PhoneAttributeTests.cs (hunk 5 above)
Comment (hughbe) [inline src/System.ComponentModel.Annotations/tests/PhoneAttributeTests.cs]: I wanted to see what kind of invalid phone numbers/false positives PhoneAttribute would accept!

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttribute.cs (hunk 2 above)
Comment (stephentoub) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttribute.cs]: I see. Ok, thanks.
